### PR TITLE
uplink/piecestore: handle Download errors better

### PR DESCRIPTION
**Pull request by egonelbre**
What: Handle errors better such that we can call close without context cancellation problems. This should also get rid of the potential double closing of download.

Why:

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: uplink/piecestore/download.go [hunk 1]
```diff
@@ -100,17 +104,25 @@ func (client *Client) Download(ctx context.Context, limit *pb.OrderLimit, pieceP
...
-		// if we have an error or are pending for an error, avoid further communication
-		// however we should still finish reading the unread data.
-		if err != nil || client.unread.Errored() {
+		// if we have an error return the error
+		if err != nil {
 			return read, err
 		}
+		// if we are pending for an error, avoid further requests, but try to finish what's in unread buffer.
+		if client.unread.Errored() {
+			return read, nil
```
Comment (mobyvb) [inline uplink/piecestore/download.go]: why don't you return an error in this case?

Comment (egonelbre) [reply uplink/piecestore/download.go]: Because `unread` still contains valid data, that can be read until it hits the error.

Comment (mobyvb) [reply uplink/piecestore/download.go]: Maybe I am just misunderstanding then. I thought if `client.unread.Errored()` is true, the error was already hit. I also thought that `Read` should return an error, even when the number of bytes read is larger than 0.

Comment (egonelbre) [reply uplink/piecestore/download.go]:
```
Let's say you have this stream of data buffered, which ends in error

|------|----------| <- `unread` contains this
. . . . . . . . . . . . . . E
|------| <-- `read` contains this

Even though unread contains an error, the reading doesn't end in the error...
So we shouldn't return the error until we have read all the valid data.
```

Comment (mobyvb) [reply uplink/piecestore/download.go]: Ok, that makes sense. It looks like in some cases, errors are added to `client.unread` which are never returned or added to `client.closingError`. And this block seems to be the only place where `client.unread.Errored()` is used except for calls to `client.unread.IncludeError()`. So I am assuming it is okay if some unread errors are not propagated outwards?

**Comment by egonelbre** [comment]
@zeebo ideally no, however it is called concurrently... but, there's a `piecestore.LockingDownload` that wraps it and handles the locking.

File: uplink/piecestore/download.go (hunk 1 above)
Comment (egonelbre) [reply uplink/piecestore/download.go]: `unread.Read` returns the error when it reads it.

Comment (mobyvb) [reply uplink/piecestore/download.go]: !!
makes sense

**Comment by mobyvb** [comment]
@zeebo @egonelbre `lazyPieceReader` also has a mutex that locks Read/Close

**Comment by egonelbre** [comment]
@zeebo yes.

**Comment by mobyvb** [comment]
I think the `TestUploadDownloadBandwidth` test is failing for a similar reason as it was failing in #2757 - not enough time passes after the test calls `Download()` for all the storagenodes to save their orders before we calculate totals. So not all the orders are being counted and the assertion fails.

**Pull request merged**